File: PalindromePartitioning.java
// Time complexity: O(2 ^ n)
// Space complexity: O(n) Recursive stack

// Approach: Using pivot from 0 to length - 1;
// Index from pivot to length - 1; Checking if substring
// from pivot to index is valid; if it is; move pivot to the index
// after the valid substring.

import java.util.ArrayList;
import java.util.List;

class PalindromePartitioning {
    List<List<String>> res;

    public List<List<String>> partition(String s) {
        this.res = new ArrayList();
        dfs(0, s, new ArrayList());
        return res;
    }

    private void dfs(int pivot, String s, List<String> path) {
        if (pivot == s.length()) {
            res.add(new ArrayList(path));
            return;
        }

        // out of bounds should be handled
        for (int i = pivot; i < s.length(); i++) {
            String sub = s.substring(pivot, i + 1);
            if (!isPalindrome(sub)) {
                continue;
            }
            path.add(sub);
            dfs(i + 1, s, path);
            path.remove(path.size() - 1);
        }
    }

    private boolean isPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;

        while (i < j) {
            if (s.charAt(i) != s.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }

        return true;
    }
}

File: Palindrome Partitioning.cpp
//Time Complexity- O(n*2^n)
//Space Complexity- O(n)

class Solution {
public:
    vector<vector<string>> partition(string s) {
        
        vector<vector<string>> ans;
        vector<string> temp;
        helper(ans,temp,s);
        return ans;
    }
    
    void helper(vector<vector<string>> &ans,vector<string> temp,string s){
        
        if(s.size()==0){
            ans.push_back(temp);
            return;
        }
        
        for(int i=0;i<s.size();i++){
            if(isPalindrome(s,0,i+1)){
                temp.push_back(s.substr(0,i+1));
                helper(ans,temp,s.substr(i+1,s.size()));
                temp.pop_back();
            }
        }
    }
    
    bool isPalindrome(string s,int start,int end){
        
        end--;
        while(start<=end){
            if(s[start]!=s[end]){
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
};